File: guild_of_heroes/guild_of_heroes_core/src/guild.cpp
#include <stdexcept>
#include "guild.hpp"

Int
guild_of_heroes::Guild::getHeroesCount() const
{
	return heroes.size();
}

std::vector<guild_of_heroes::Hero*>
guild_of_heroes::Guild::getHeroes() const
{
	return heroes;
}

void
guild_of_heroes::Guild::addHero(Hero* hero)
{
	if (hero == nullptr)
		throw std::invalid_argument("Hero can not be null");
	heroes.push_back(hero);
}

guild_of_heroes::Guild*
guild_of_heroes::Guild::create(const std::string& name) {
	if (name.empty())
		throw std::invalid_argument("name can not be empty");
	return new Guild(name);
}

guild_of_heroes::Guild::Guild(const std::string& name)
	: Nameable(name), heroes()
{ }

File: GuildOfHeroes/GuildOfHeroes.Core/Guild.cs
using GuildOfHeroes.Core.Base;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace GuildOfHeroes.Core
{
    public class Guild : NameableBase
    {
        public int HeroesCount => heroes.Count;
        public IReadOnlyList<Hero> Heroes => heroes;
        public int HeroesDailyPayment => heroes.Sum(h => h.DailyFee);

        public void AddHero(Hero hero)
        {
            if (hero == null)
                throw new ArgumentException("Hero can not be null");            
            if (heroes.Contains(hero))
                throw new ArgumentException("Hero already added");

            heroes.Add(hero);
        }
        public void RemoveHero(Hero hero)
        {
            if(!heroes.Remove(hero))
                throw new InvalidOperationException(
                    "Attempt to delete non-existent hero"
                );
        }
        public void AddResource(string name, int count)
        {

        }
        public int GetResourcesTypesCount()
        {
            return 0;
        }


        public static Guild Create(string name)
        {
            if (name.Length == 0)
                throw new ArgumentException("name can not be empty");
            return new Guild(name);
        }



        private Guild(string name)
            : base(name)
        {
            heroes = new List<Hero>();
        }

        private List<Hero> heroes;
    }
}
